### PR TITLE
Using the wrong protocol to pull from ghcr.io gives non-intuitive errors

### LINKED ISSUE #632: Using the wrong protocol to pull from ghcr.io gives non-intuitive errors
ghcr.io supports both `docker://` and `oras://` protocols, but with at least singularity-3.8.4 and later, attempting to pull a container using the wrong protocol gives non-intuitive answers.  For example with apptainer-1.1.0-rc.1:
```
$ apptainer pull oras://ghcr.io/apptainer/lolcow:latest
FATAL:   While pulling image from oci registry: error fetching image to cache: failed to get checksum for oras://ghcr.io/apptainer/lolcow:latest: could not get image manifest, received mediaType: application/vnd.docker.distribution.manifest.v2+json
$ apptainer pull docker://ghcr.io/apptainer/lolcow:latest
...
INFO:    Creating SIF file...
$ apptainer pull docker://ghcr.io/apptainer/alpine
...
FATAL:   While making image from oci registry: error fetching image to cache: while building SIF from layers: conveyor failed to get: unsupported image-specific operation on artifact with type "application/vnd.unknown.config.v1+json"
$ apptainer pull oras://ghcr.io/apptainer/alpine:latest
INFO:    Downloading oras image
```
It would be nice if the message would be more helpful and give a hint to switch protocols.

### PR DESCRIPTION
**Pull request by JasonYangShadow**
Signed-off-by: jason yang <jasonyangshadow@gmail.com>

## Description of the Pull Request (PR):

Change the error message to provide better intuitive errors.

### This fixes or addresses the following GitHub issues:

 - Fixes #632 

#### Before submitting a PR, make sure you have done the following:

- Read the [Guidelines for Contributing](https://github.com/apptainer/apptainer/blob/main/CONTRIBUTING.md), and this PR conforms to the stated requirements.
- Added changes to the [CHANGELOG](https://github.com/apptainer/apptainer/blob/main/CHANGELOG.md) if necessary according to the [Contribution Guidelines](https://github.com/apptainer/apptainer/blob/main/CONTRIBUTING.md)
- Added tests to validate this PR, linted with `make check`  and tested this PR locally with a `make test`, and `make testall` if possible (see CONTRIBUTING.md).
- Based this PR against the appropriate branch according to the [Contribution Guidelines](https://github.com/apptainer/apptainer/blob/main/CONTRIBUTING.md)
- Added myself as a contributor to the [Contributors File](https://github.com/apptainer/apptainer/blob/main/CONTRIBUTORS.md)

### CODE REVIEW

File: internal/pkg/client/oci/pull.go [hunk 1]
```diff
@@ -108,6 +109,9 @@ func PullToFile(ctx context.Context, imgCache *cache.Handle, pullTo, pullFrom, t
 
 	src, err := pull(ctx, imgCache, directTo, pullFrom, tmpDir, ociAuth, noHTTPS, noCleanUp)
 	if err != nil {
+		if strings.Contains(err.Error(), "unsupported image-specific operation on artifact with type \"application/vnd.unknown.config.v1+json\"") {
+			return "", fmt.Errorf("while building SIF from layers: %v, you might need to change the protocol to oras://", err)
```
Comment (DrDaveD) [inline internal/pkg/client/oci/pull.go:113]: I think the "while building SIF from layers" is supposed to be added at a higher level, not in this function.  Change it to just "%v; try changing the protocol to oras://".

File: internal/pkg/client/oras/oras.go [hunk 2]
```diff
@@ -285,6 +287,9 @@ func ImageSHA(ctx context.Context, uri string, ociAuth *ocitypes.DockerAuthConfi
 
 	// ensure that we received an image manifest descriptor
 	if desc.MediaType != ocispec.MediaTypeImageManifest {
+		if desc.MediaType == manifest.DockerV2Schema2MediaType {
+			return "", errors.New("please use docker:// instead of oras:// to pull the image")
```
Comment (DrDaveD) [inline internal/pkg/client/oras/oras.go:291]: This message should say something about an error before there's a suggestion of what to do.  Start it with "unexpected docker media type received; "

**Review by DrDaveD — CHANGES_REQUESTED** [review]

File: internal/pkg/client/oci/pull.go (hunk 1 above)
Comment (JasonYangShadow) [reply internal/pkg/client/oci/pull.go]: thanks for the comment! reflected in this commit https://github.com/apptainer/apptainer/pull/703/commits/ed8c9af19331d20eb636b00bb05d4a7517f03ac8

File: internal/pkg/client/oras/oras.go (hunk 2 above)
Comment (JasonYangShadow) [reply internal/pkg/client/oras/oras.go]: thanks for the comment! reflected in this commit https://github.com/apptainer/apptainer/pull/703/commits/ed8c9af19331d20eb636b00bb05d4a7517f03ac8

File: internal/pkg/client/oci/pull.go [hunk 3]
```diff
@@ -108,6 +109,9 @@ func PullToFile(ctx context.Context, imgCache *cache.Handle, pullTo, pullFrom, t
 
 	src, err := pull(ctx, imgCache, directTo, pullFrom, tmpDir, ociAuth, noHTTPS, noCleanUp)
 	if err != nil {
+		if strings.Contains(err.Error(), "unsupported image-specific operation on artifact with type \"application/vnd.unknown.config.v1+json\"") {
+			return "", fmt.Errorf("%v;try changing the protocol to oras://", err)
```
Comment (DrDaveD) [inline internal/pkg/client/oci/pull.go:113]: Please put a blank after the semicolon.

File: internal/pkg/client/oras/oras.go [hunk 4]
```diff
@@ -285,6 +287,9 @@ func ImageSHA(ctx context.Context, uri string, ociAuth *ocitypes.DockerAuthConfi
 
 	// ensure that we received an image manifest descriptor
 	if desc.MediaType != ocispec.MediaTypeImageManifest {
+		if desc.MediaType == manifest.DockerV2Schema2MediaType {
+			return "", errors.New("unexpected docker media type received;try changing the protocol to docker://")
```
Comment (DrDaveD) [inline internal/pkg/client/oras/oras.go:291]: Also a blank after the semicolon here, please.

**Review by DrDaveD — CHANGES_REQUESTED** [review]

**Comment by DrDaveD** [comment]
Thanks.  Please cherry-pick this to a PR on the release-1.1 branch as well.

**Pull request merged**